File: Simplex.Core/Actions/IActionBinderSet.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Text;

namespace Tmon.Simplex.Actions
{
    /// <summary>
    /// 액션 래핑 클래스를 그룹화형 관리할 인터페이스
    /// </summary>
    public interface IActionBinderSet
    {
        /// <summary>
        /// 액션 래핑 클래스를 조회 (등록된 액션 래핑 클래스가 존재하지 않으면  신규로 등록 함)
        /// </summary>
        /// <typeparam name="TActionBinderSet"></typeparam>
        /// <returns></returns>
        TActionBinderSet GetOrAdd<TActionBinderSet>()
            where TActionBinderSet : IActionBinderSet, new();

        /// <summary>
        /// 단위작업 바인더를 조회합니다. 단위작업이 등록되어 있지 않으면 새롭게 등록하고 해당 객체를 리턴합니다.
        /// </summary>
        /// <typeparam name="TAction">단위작업의 타입</typeparam>
        /// <typeparam name="TParam">단위작업의 매개변수 타입</typeparam>
        /// <typeparam name="TResult">단위작업의 결과 반환 타입</typeparam>
        /// <param name="key">단위작업의 등록 키</param>
        /// <returns>단위작업 바인더 인스턴스</returns>
        IActionBinder<TAction, TParam, TResult> GetOrAdd<TAction, TParam, TResult>([CallerMemberName] string actionBinderKey = "")
           where TAction : class, IAction<TParam, TResult>, new();

        /// <summary>
        /// 단위작업 바인더를 조회합니다. 단위작업이 등록되어 있지 않으면 새롭게 등록하고 해당 객체를 리턴합니다.
        /// </summary>
        /// <typeparam name="TAction">단위작업의 타입</typeparam>
        /// <typeparam name="TResult">단위작업의 결과 반환타입</typeparam>
        /// <param name="key">단위작업의 등록 키</param>
        /// <returns>단위작업 바인더 인스턴스</returns>
        IActionBinder<TAction, TResult> GetOrAdd<TAction, TResult>([CallerMemberName] string actionBinderKey = "")
            where TAction : class, IAction<TResult>, new();
    }
}

File: simplex.java/src/main/java/kr/co/tmon/simplex/actions/IActionBinderSet.java
package kr.co.tmon.simplex.actions;

/**
 * 액션 클래스의 명세를 그룹으로 관리할 인터페이스
 * @author yookjy
 *
 */
public interface IActionBinderSet {
	
	/**
	 * 액션 그룹 명세를 조회 (액션 그룹이 등록되지 않았으면  신규로 등록 후 리턴)
	 * @param clazz 액션 그룹 타입
	 * @return 액션 그룹
	 * @throws InstantiationException
	 * @throws IllegalAccessException
	 */
	<TActionBinderSet extends IActionBinderSet> TActionBinderSet getOrAdd(Class<TActionBinderSet> clazz) 
			throws InstantiationException, IllegalAccessException;
	
	/**
	 * 액션 클래스 명세를 조회 (액션 클래스의 명세가 등록되지 않았으면 신규로 등록 후 리턴)
	 * @param actionBinderKey 액션 클래스의 명세 키
	 * @param clazz 액션 클래스의 타입
	 * @return 액션 클래스 명세
	 */
	<TAction extends IAction<TParam, TResult>, TParam, TResult> 
		IActionBinder<TAction, TParam, TResult> getOrAdd(String actionBinderKey, Class<TAction> clazz);
}
